contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,251 @@
+# Watchable API Group Discovery
+
+The Kubernetes API server provides a number of HTTP endpoinds to
+discover the available API groups, versions and resources:
+
+- `/apis`: returns a list of groups with available versions (compare [metav1.APIGroupList](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L639))
+- `/apis/<group>`: returns a list of versions, **without** resources (compare [metav1.APIVersions](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L623))
+- `/apis/<group>/<version>`: returns a list of resources (compare [metav1.APIResourceList](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L718))
+
+This discovery information is used by many API clients like kubectl
+and controllers like the namespace controller. The garbage
+collection controller is supposed to use discovery as well, [but
+doesn't yet](https://github.com/kubernetes/kubernetes/blob/6f5337cff7baffc2b27d85036cb1ecb6a4c1f73b/cmd/kube-controller-manager/app/core.go#L174).
+
+Kubectl will cache the discovery information in
+`~/.kube/cache` for ten minutes and queries the API server again after
+that time or on a cache miss on RESTMapper lookups. In the namespace controller
```

Comment:
Can be a misspelled resource name, can be a new TPR name. Otherwise, the 10 min apply.